### DISCUSSION TITLE
How to properly insert an image in PDF with PyMuPDF 1.22.*?

### DISCUSSION THREAD

**Discussion by ipoyt** [pymupdf/PyMuPDF#3497] [Looking for help] [answered]
Hello!
I have run into a little problem. I need to paste a new image in a PDF file which isn't stored or saved there; I have a base46-string image `img`:

```
imgdata = base64.b64decode(img)
im = io.BytesIO(pngdata)
```

Then I used the guide from [PyMuPDF insert image at bottom](https://stackoverflow.com/questions/60711402/pymupdf-insert-image-at-bottom):

```
if not page.is_wrapped:
    page.wrap_contents()
page._insert_image(rect, stream=im)
```

Firstly, there's no `insertImage` metod available in 1.22.*, I found `_insert_image` instead.
Secondly, the output is:
```
return _fitz.Page__insert_image(self, filename, pixmap, stream, imask, clip, overlay, rotate, keep_proportion, oc, width, height, xref, alpha, _imgname, digests)
TypeError: in method 'Page__insert_image', argument 2 of type 'char *'
```

Do you have any ideas what I am doing wrong?

**Comment by JorjMcKie** [comment, marked as answer]
The image insert method is [insert_image](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_image).
You normally do not need to convert the binary data output `imgdata` from base64 via PIL first. That is to say, all image formats complying with https://pymupdf.readthedocs.io/en/latest/pixmap.html#supported-input-image-formats can be used directly as values for parameter `stream`.

**Reply by ipoyt** [reply]
I see, thank you!
